### ISSUE TITLE
SAX parser silently exits on big files

### ISSUE THREAD

**Issue by ivan-kleshnin** [inikulin/parse5#97]
The problem seems to involve only SAX parser.

#### Basic code example

```js
import Fs from "fs";
import Parse5 from "parse5";

process.on("unhandledRejection", function (reason, p) {
  throw reason;
});

let source = Fs.createReadStream("heavy.html");

let parser = new Parse5.SAXParser();
let pCount = 0;

parser.on("startTag", function (name, attrs) {
  if (name == "p") {
    pCount += 1;
    console.log("p #" + pCount);
  }
});

parser.on("finish", function () {
  console.log("Total pCount =", pCount);
});

source.pipe(parser);
```

As `heavy.html` starts to be bigger than **~64Kib** SAX parser exits silently.
`finish` error never runs than.

If `heavy.html` is smaller, everything works fine. 
Both `startTag` and `finish` event run in expected way.

**Node** version: v4.2.2
**Parse5** version: 2.1.1

---

##### Example file
http://paqmind.com/heavy.html

##### Output

```
p #1
p #2
p #3
p #4
p #5
p #6
p #7
p #8
p #9
p #10
p #11
p #12
p #13
p #14
p #15
p #16
p #17
p #18
p #19
p #20
p #21
p #22
p #23
p #24
p #25
p #26
p #27
p #28
p #29
p #30
p #31
p #32
p #33
p #34
p #35
p #36
p #37
p #38
p #39
p #40
p #41
p #42
p #43
p #44
p #45
p #46
p #47
p #48
p #49
p #50
p #51
p #52
p #53
p #54
p #55
p #56
p #57
p #58
p #59
p #60
p #61
p #62
p #63
p #64
p #65
p #66
```

Remove some paragraphs from "heavy.html" and `finish` event starts to fire... :disappointed:

**Comment by inikulin** [comment]
Ok, I was able to localize the issue. `TransformStream` designed the way that if it's `ReadableStream` part doesn't have consumers it's stops piping data through `_transform` once it's reached `highWaterMark` (which is 16Kb by default). The first chunk of this file already exceeds this limit, so the remaining data will not be parsed. But once we have consumer - everything works just fine, you can try:
```js
source.pipe(parser).pipe(process.stdout);
```
Seems like setting `highWaterMark` to `Number.MAX_VALUE` in `TransformStream` ctor should do a trick (since we need to consume whole data anyway). But I wonder if there are any caveats. It will be nice to have a feedback from some node streams expert.

**Comment by inikulin** [comment]
Nevermind, I found workaround :wink:

**Comment by domenic** [comment]
In general the point of a highWaterMark is to prevent more than that amount of memory from being consumed. So as long as the workaround sticks to that constraint, it is in line with the spirit of streams.

**Comment by inikulin** [comment]
@domenic Yep, I just pipe to /dev/null stream by default, so I can keep default `highWaterMark` value, but let data flow through transform if we don't have consumers

**Comment by inikulin** [comment]
Published as 2.1.2
